<a href="https://colab.research.google.com/github/priyanshgupta1998/Machine_learning/blob/master/Quora_Scraping/Part1_Scraping_Data/Quora_Scraping.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [0]:
# import pandas as pd
# from urllib.request import urlopen
# from bs4 import BeautifulSoup

In [0]:
# url = "https://www.google.com/search?q=titan+watches+quora+site:www.quora.com&client=ms-android-xiaomi&prmd=imnv&sxsrf=ACYBGNS1ycXj7ojWUEc5BzNpwOeGr9555w:1568619197468&sa=X&ved=2ahUKEwjY7LXw6dTkAhWJxMQBHXd9DB0QrQIoBDAAegQIAxAK&biw=360&bih=615#ip=1"

# html = urlopen(url)

In [0]:
# from urllib.parse import urlencode, urlparse, parse_qs

# from lxml.html import fromstring
# from requests import get

# raw = get(url)
# raw = get(url).text
# raw

# page = fromstring(raw)
# page

# list(page)



---



---



---



---



In [0]:
# url = "https://www.google.com/search?q=titan+watches+quora+site:www.quora.com&client=ms-android-xiaomi&prmd=imnv&sxsrf=ACYBGNS1ycXj7ojWUEc5BzNpwOeGr9555w:1568619197468&sa=X&ved=2ahUKEwjY7LXw6dTkAhWJxMQBHXd9DB0QrQIoBDAAegQIAxAK&biw=360&bih=615#ip=1"


# from urllib.request import urlopen as ureq
# from bs4 import BeautifulSoup
# import requests
# page = requests.get(url)
# html = BeautifulSoup(page.content,'html.parser')
# html

In [0]:
# div1 = html.findAll("a")

In [0]:
# for link in div1:
#   if('/url?q=' in link.get('href')):
#     if('https://www.quora.com/' in link.get('href').split('/url?q=')[1].split('&sa=')[0]):
#       print(link.get('href').split('/url?q=')[1].split('&sa=')[0])

https://www.quora.com/How-good-are-titan-watches-compared-to-other-global-brands
https://www.quora.com/What-is-your-opinion-on-Titan-watches
https://www.quora.com/Why-are-Titan-watches-so-expensive
https://www.quora.com/Which-watch-brand-should-I-go-for-Titan-Fossil-or-Casio
https://www.quora.com/How-is-the-quality-of-Sonata-s-watches-Is-it-as-good-as-Titan-s-watches
https://www.quora.com/How-popular-are-Titan-Watches-outside-India
https://www.quora.com/Is-it-worth-buying-a-Fossil-wrist-watch-instead-of-Titan-Timex-or-Sonata-and-why
https://www.quora.com/Which-are-the-best-models-of-Titan-watches-for-men
https://www.quora.com/Which-watch-brand-makes-durable-and-long-life-watches-Titan-Maxima-or-Casio
https://www.quora.com/Which-watch-is-better-Titan-or-Casio




---



---



---



---



---

#finally we are goona scrape 150 Google Queries for watch Searching on Quora 

In [1]:
link ='https://www.google.com/search?q=titan+watches+quora+site:www.quora.com&client=ms-android-xiaomi&biw=360&bih=615&sxsrf=ACYBGNQS6NzTZhe1c4WSZCqqSuZxxZUzqg:1570278410182&ei=CoyYXczXCo2e9QOBqqzIBQ&start=0&sa=N&ved=0ahUKEwjMhNn2joXlAhUNT30KHQEVC1kQ8tMDCKgB'

links= []
k = link.split('&start=0')
for i in range(15):   #I've taken 15 Google pages here 
  links.append(k[0]+'&start='+str(i*10)+k[1])

print(len(links))
from urllib.request import urlopen as ureq
from bs4 import BeautifulSoup
import requests
total =[]
for link in links:
  page = requests.get(link)
  html = BeautifulSoup(page.content,'html.parser')
  div1 = html.findAll("a")
  for link in div1:
    if('/url?q=' in link.get('href')):
      if('https://www.quora.com/' in link.get('href').split('/url?q=')[1].split('&sa=')[0]):
        total.append(link.get('href').split('/url?q=')[1].split('&sa=')[0])
        
print(len(total))
total[:15]

15
150


['https://www.quora.com/How-good-are-titan-watches-compared-to-other-global-brands',
 'https://www.quora.com/What-is-your-opinion-on-Titan-watches',
 'https://www.quora.com/Why-are-Titan-watches-so-expensive',
 'https://www.quora.com/Which-watch-brand-should-I-go-for-Titan-Fossil-or-Casio',
 'https://www.quora.com/How-is-the-quality-of-Sonata-s-watches-Is-it-as-good-as-Titan-s-watches',
 'https://www.quora.com/How-popular-are-Titan-Watches-outside-India',
 'https://www.quora.com/Is-it-worth-buying-a-Fossil-wrist-watch-instead-of-Titan-Timex-or-Sonata-and-why',
 'https://www.quora.com/Which-are-the-best-models-of-Titan-watches-for-men',
 'https://www.quora.com/Which-watch-brand-makes-durable-and-long-life-watches-Titan-Maxima-or-Casio',
 'https://www.quora.com/Which-watch-is-better-Titan-or-Casio',
 'https://www.quora.com/Which-one-is-best-Titan-or-Fastrack',
 'https://www.quora.com/Why-are-Titan-watches-coveted-among-collectors',
 'https://www.quora.com/For-how-many-years-will-a-Titan-

In [0]:
len(set(total))    # So we can say all the links are completely different

150

#Save all the links in csv file

In [0]:
import pandas as pd
df= pd.DataFrame()
df['150_quora_Links'] = total
df.to_csv('Links_file.csv' , index =False)

#Now take start taking all the links one by one in order to scrape

In [0]:
# from urllib.request import urlopen as ureq
# from bs4 import BeautifulSoup
# import requests
# link = 'https://www.quora.com/How-good-are-titan-watches-compared-to-other-global-brands'
# page = requests.get(link)
# html = BeautifulSoup(page.content,'html.parser')
# div1 = html.findAll('div',{'class':'pagedlist_item'})
# # print(div1[0].findAll('a' , {'class':'user'}))
# print(len(div1))
# # print(div1[0].prettify())

In [0]:
# from urllib.request import urlopen as ureq
# from bs4 import BeautifulSoup
# import requests
# for link in total[:5]:
#   page = requests.get(link)
#   html = BeautifulSoup(page.content,'html.parser')
#   div1 = html.findAll('div',{'class':'pagedlist_item'})
#   for i in range(len(div1)):
#     if(len(div1[i].text)>0): 
#       #ReadingContentFooter AnswerFooter ContentFooter
#       #AnswerFooter ContentFooter ReadingContentFooter
#       #AnswerFooter ReadingContentFooter ContentFooter
#       if(len(div1[i].findAll('a' , {'class':'user'}))>0):
#         print(div1[i].findAll('a' , {'class':'user'})[0].text)
#       else:
#         print("^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^")
#       view1 = div1[i].findAll('div' , {'class':'AnswerFooter ContentFooter ReadingContentFooter'})
#       view2 = div1[i].findAll('div' , {'class':'AnswerFooter ReadingContentFooter ContentFooter'})
#       view3 = div1[i].findAll('div' , {'class':'ReadingContentFooter AnswerFooter ContentFooter'})
#       if(len(view1)==0):
#         if(len(view2)==0):
#           if(len(view3)>0):
#             st=view3[0].text
#         else:
#           st=view2[0].text
#       else:
#         st=view1[0].text
#       if('views' in st):
#         print(st.split('·')[0])
#       if('voter' in st):
#         print(st.split('·')[1])
#       print("+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n")
#     else:
#       print("+++++++++++++++++++++++++++++++++++++++++GO ERROR ++++++++++++++++++++++++++++++++++++++++++++\n")
#   print("***************************************Next CHnage***********************\n")

#ALL data about First 150 Relevant Questions to Watches on Quora.com

#Scrape first 5 appearing Questions for this search

In [1]:
import pandas as pd
links = pd.read_csv('Links_file.csv')
total = list(links['150_quora_Links'])
total[:5]

['https://www.quora.com/How-good-are-titan-watches-compared-to-other-global-brands',
 'https://www.quora.com/What-is-your-opinion-on-Titan-watches',
 'https://www.quora.com/Why-are-Titan-watches-so-expensive',
 'https://www.quora.com/Which-watch-brand-should-I-go-for-Titan-Fossil-or-Casio',
 'https://www.quora.com/How-is-the-quality-of-Sonata-s-watches-Is-it-as-good-as-Titan-s-watches']

#I'm removing 20 Google Queries .because I don't have Enough RAM . and these 20 Quesies (below) are taking too much RAM to scrape the data

In [2]:
joker = total
del joker[38]
del joker[40]
del joker[43]
del joker[43]
del joker[62]
del joker[79]
del joker[79]
del joker[82]
del joker[89]
del joker[95]
del joker[118]
del joker[126]
del joker[136]
del joker[82]
del joker[97]
del joker[98]
del joker[108]
del joker[108]
del joker[127]
del joker[62]
print(len(joker))

130


#So i'm taking only 130 Google Queries

#Let's Scrape the data Manually using Beautiful Soup Library

In [5]:
import time
from urllib.request import urlopen as ureq
from bs4 import BeautifulSoup
import requests
user = []
background = []
Answer = []
date = []
urlss = []
question = []
num_of_ans = []
viewers =[]
updates = []
for link in joker:
  page = requests.get(link)
  html = BeautifulSoup(page.content,'html.parser')
  ques = html.findAll('h1')
  ans =html.findAll('div',{'class':'QuestionPageAnswerHeader'})
  li= int(ans[0].text.split('Answ')[0])
  div1 = html.findAll('div',{'class':'pagedlist_item'})
  print(joker.index(link) , link , len(div1))
  for i in range(len(div1)):
    if(len(div1[i].text)>0):
      urlss.append(link)
      question.append(ques[0].text)
      num_of_ans.append(li)
      if(len(div1[i].findAll('a' , {'class':'user'}))>0):
        user.append(div1[i].findAll('a' , {'class':'user'})[0].text)
      else:
        user.append('?')
      st =''  
      view1 = div1[i].findAll('div' , {'class':'AnswerFooter ContentFooter ReadingContentFooter'})
      view2 = div1[i].findAll('div' , {'class':'AnswerFooter ReadingContentFooter ContentFooter'})
      view3 = div1[i].findAll('div' , {'class':'ReadingContentFooter AnswerFooter ContentFooter'})
      if(len(view1)==0):
        if(len(view2)==0):
          if(len(view3)>0):
            st=view3[0].text
        else:
          st=view2[0].text
      else:
        st=view1[0].text

      if('views' in st):
        viewers.append(st.split('·')[0])
      else:
        viewers.append('?')
      if('voter' in st):
        updates.append(st.split('·')[1])
      else:
        updates.append('?')
        
        
      if(len(div1[i].findAll('span' , {'class':'NameCredential'}))>0):
        background.append(div1[i].findAll('span' , {'class':'NameCredential'})[0].text)
      else:
        background.append('?')

      #print(div1[i].findAll('span' , {'class':'bold_num'}))
      s = ' '
      k = div1[i].findAll('p',{'class':"ui_qtext_para u-ltr u-text-align--start"})  #Answer of the asked question
      for j in range(len(k)):
        s+= k[j].text 
        s+=s
      Answer.append(s)

      if(len(div1[i].findAll('a' , {'class':'answer_permalink'}))>0):
        date.append(div1[i].findAll('a' , {'class':'answer_permalink'})[0].text)
      else:
        date.append('?')
#       if(len(div1[i].findAll('div' , {'id':'wNtBhqKo28'}))>0):
#         print(div1[i].findAll('div' , {'id':'wNtBhqKo28'})[0].text)
#       else:
#         print("^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^")
#       print("+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n")
#     else:
#       print("+++++++++++++++++++++++++++++++++++++++++GO ERROR ++++++++++++++++++++++++++++++++++++++++++++\n")
#   print("***************************************Next CHnage***********************\n")

0 https://www.quora.com/How-good-are-titan-watches-compared-to-other-global-brands 20
1 https://www.quora.com/What-is-your-opinion-on-Titan-watches 9
2 https://www.quora.com/Why-are-Titan-watches-so-expensive 7
3 https://www.quora.com/Which-watch-brand-should-I-go-for-Titan-Fossil-or-Casio 20
4 https://www.quora.com/How-is-the-quality-of-Sonata-s-watches-Is-it-as-good-as-Titan-s-watches 7
5 https://www.quora.com/How-popular-are-Titan-Watches-outside-India 4
6 https://www.quora.com/Is-it-worth-buying-a-Fossil-wrist-watch-instead-of-Titan-Timex-or-Sonata-and-why 14
7 https://www.quora.com/Which-are-the-best-models-of-Titan-watches-for-men 9
8 https://www.quora.com/Which-watch-brand-makes-durable-and-long-life-watches-Titan-Maxima-or-Casio 18
9 https://www.quora.com/Which-watch-is-better-Titan-or-Casio 11
10 https://www.quora.com/Why-are-Titan-watches-coveted-among-collectors 4
11 https://www.quora.com/Which-one-is-best-Titan-or-Fastrack 7
12 https://www.quora.com/For-how-many-years-will-

#create a pandas dataframe and store all the data in a .csv file format

In [0]:
import pandas as pd
df = pd.DataFrame()
df['Question'] = question
df['Number_of_Answer'] = num_of_ans
df['User_name'] = user
df['User_Profile'] = background
df['Answer'] = Answer
df['Number_of_Views'] = viewers
df['Number_of_Upvotes'] = updates
df['Date_Of_Answer'] =date
df['Question_link'] = urlss


In [9]:
print(df.shape)
df.head()

(877, 9)


Question  ...                                      Question_link
0  How good are titan watches compared to other g...  ...  https://www.quora.com/How-good-are-titan-watch...
1  How good are titan watches compared to other g...  ...  https://www.quora.com/How-good-are-titan-watch...
2  How good are titan watches compared to other g...  ...  https://www.quora.com/How-good-are-titan-watch...
3  How good are titan watches compared to other g...  ...  https://www.quora.com/How-good-are-titan-watch...
4  How good are titan watches compared to other g...  ...  https://www.quora.com/How-good-are-titan-watch...

[5 rows x 9 columns]

In [0]:
df.to_csv('Quora_Questions_Scraping.csv',index=False)

#From above details ,we can clearly see we have 130 Questions along with total 877 Answers.

#END



---

---



---



---



---



---



---





---



---



---



In [0]:
# import time
# from urllib.request import urlopen as ureq
# from bs4 import BeautifulSoup
# import requests
# user = []
# background = []
# Answer = []
# date = []
# urlss = []
# question = []
# num_of_ans = []
# viewers =[]
# updates = []
# for link in joker[:70]:
#   page = requests.get(link)
#   html = BeautifulSoup(page.content,'html.parser')
#   ques = html.findAll('h1')
#   ans =html.findAll('div',{'class':'QuestionPageAnswerHeader'})
#   li= int(ans[0].text.split('Answ')[0])
#   div1 = html.findAll('div',{'class':'pagedlist_item'})
#   print(joker.index(link) , link , len(div1))
#   for i in range(len(div1)):
#     if(len(div1[i].text)>0):
#       urlss.append(link)
#       question.append(ques[0].text)
#       num_of_ans.append(li)
#       if(len(div1[i].findAll('a' , {'class':'user'}))>0):
#         user.append(div1[i].findAll('a' , {'class':'user'})[0].text)
#       else:
#         user.append('?')
#       st =''  
#       view1 = div1[i].findAll('div' , {'class':'AnswerFooter ContentFooter ReadingContentFooter'})
#       view2 = div1[i].findAll('div' , {'class':'AnswerFooter ReadingContentFooter ContentFooter'})
#       view3 = div1[i].findAll('div' , {'class':'ReadingContentFooter AnswerFooter ContentFooter'})
#       if(len(view1)==0):
#         if(len(view2)==0):
#           if(len(view3)>0):
#             st=view3[0].text
#         else:
#           st=view2[0].text
#       else:
#         st=view1[0].text

#       if('views' in st):
#         viewers.append(st.split('·')[0])
#       else:
#         viewers.append('?')
#       if('voter' in st):
#         updates.append(st.split('·')[1])
#       else:
#         updates.append('?')
        
        
#       if(len(div1[i].findAll('span' , {'class':'NameCredential'}))>0):
#         background.append(div1[i].findAll('span' , {'class':'NameCredential'})[0].text)
#       else:
#         background.append('?')

#       #print(div1[i].findAll('span' , {'class':'bold_num'}))
#       s = ' '
#       k = div1[i].findAll('p',{'class':"ui_qtext_para u-ltr u-text-align--start"})  #Answer of the asked question
#       for j in range(len(k)):
#         s+= k[j].text 
#         s+=s
#       Answer.append(s)

#       if(len(div1[i].findAll('a' , {'class':'answer_permalink'}))>0):
#         date.append(div1[i].findAll('a' , {'class':'answer_permalink'})[0].text)
#       else:
#         date.append('?')
# #       if(len(div1[i].findAll('div' , {'id':'wNtBhqKo28'}))>0):
# #         print(div1[i].findAll('div' , {'id':'wNtBhqKo28'})[0].text)
# #       else:
# #         print("^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^")
# #       print("+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n")
# #     else:
# #       print("+++++++++++++++++++++++++++++++++++++++++GO ERROR ++++++++++++++++++++++++++++++++++++++++++++\n")
# #   print("***************************************Next CHnage***********************\n")

In [0]:
# import pandas as pd
# df = pd.DataFrame()
# df['Question'] = question
# df['Number_of_Answer'] = num_of_ans
# df['User_name'] = user
# df['User_Profile'] = background
# df['Answer'] = Answer
# df['Number_of_Views'] = viewers
# df['Number_of_Upvotes'] = updates
# df['Date_Of_Answer'] =date
# df['Question_link'] = urlss

In [0]:
# print(df.shape)
# df.head()

In [0]:
# df.to_csv('QuoraScrape_SPLIT_1.csv',index=False)

In [0]:
# print(div1[2].findAll('a' , {'class':'user'}))
# print(div1[0].findAll('span' , {'class':'NameCredential'})[0].text)
# print(div1[0].findAll('span' , {'class':'bold_num'}))
# k = div1[0].findAll('p',{'class':"ui_qtext_para u-ltr u-text-align--start"})  #Answer of the asked question
# for i in range(len(k)):
#   print(k[i].text, end=' ')
# print('')
# print(div1[0].findAll('a' , {'class':'answer_permalink'})[0].text)
# print(div1[0].findAll('div' , {'id':'wNtBhqKo28'})[0].text)

In [0]:
# from scrapy import Selector
# import requests
# url = 'https://www.quora.com/How-good-are-titan-watches-compared-to-other-global-brands'
# html = requests.get(url).content
# sel = Selector( text =html )
# ass = sel.css( 'div.pagedlist_item' ).extract()
# print(len(ass))
# # print(len(sel.xpath('//div[@class="pagedlist_item"]') ))
# # sel.xpath('//div[@class="pagedlist_item"]')[0].extract()